### ISSUE TITLE
Whats the easiest way to preprocess and use the dataset for a relation extraction (RE) fine tuning task?

### ISSUE THREAD

**Issue by jambo6** [sujunhao/RENET2#3]
Hello,

I'm really interested to have a go with this dataset as the RE training dataset for something like biobert. If you look at the biobert models they use GAD or euadpr, both of which upon closer inspection are not very good. 

The ideal format is for any gene/disease combination convert mentions in the abstract/fulltext to @GENE$, @DISEASE$. So that sentences will look like:

> These results suggest that the C1772T polymorphism in @GENE$ is not involved in progression or metastasis of @DISEASE$

Training would happen on these replaced examples and performance ascertained on the test set. Is doing something like this possible using the repo functionality? I'm not managing to get the preprocessing scripts to work very well and cant tell what exactly the preprocessed format is supposed to be. Any help would be appreciated. 

Also, an unrelated question, when analysing all the PMC articles/PMID abstracts to generate the `gda_rst` files, how did you 

1. Extract the gene/disease names from the texts.
2. Convert the gene/disease names to corresponding medical ids (e.g. MESH).

Sorry for the long question(s), thanks!

**Comment by sujunhao** [comment]
Hi,

Thank you so much for being interested in using our dataset.

As you mentioned, in order to feed a document into BioBERT for finetuning, we must retrieve the document's text and replace a pair of entities into @gene$ and @disease$ at a time. This information is contained in the files docs.txt/anns.txt/labels.txt of our dataset.

To integrate this data into BioBERT, we'll need a script to do so. I'm sorry that we don't currently have a script that is specifically designed for integrating BioBERT format data. As a result, you may need to write a  integration script for BioBERT finetuning.

By the way, finetuning BioBERT with an entire document is not trivial, as BioBERT and BERT limit the number of input tokens to 512. (In other words, while an abstract may be able to feed BioBERT, the full text is completely incompatible). Therefore, an alternative method of feeding a document into BERT is to cut or filter the context into 512 tokens or to use document-level BERT such as docBERT, BERT-AL, or similar.


To answer the unrelated question: 
1. we are using the [PTC API](https://www.ncbi.nlm.nih.gov/research/pubtator/api.html) to gather information about the entities.
2. and the PTC API contains both entities and their associated IDs.

and we included a parsing [script](https://github.com/sujunhao/RENET2#pipeline-use-renet2-to-predict-gene-disease-associations-from-articles-id) that converted the text from the API to the RENET2 input format (you may need to download all data first before running the example), for your reference, our processing pipeline is like:

> input: PMID or PMCID
> output: Gene-Disease Assoications 
> step 1: download annotation from PTC API, [example](https://github.com/sujunhao/RENET2#pipeline-use-renet2-to-predict-gene-disease-associations-from-articles-id).
> step 2: parsed the downloaded annotation into RENET2 format docs.txt, sentences.txt, anns.txt)
> step 3: in RENET2 parse the RENET2 format dataset into RENET2 model input as in [this line](https://github.com/sujunhao/RENET2/blob/55c4f97883fa80d24decd96176f1a659acba12cb/src/renet2/predict.py#L48) to get tensor input
> step 4: feed the tensor input into RENET2 model and get Gene-Disease Assoications

**Comment by jambo6** [comment]
Awesome, this makes perfect sense. I've got this to work using Microsoft's BiomedNLP-PubMedBERT and it seems to work pretty well. 

I noticed that all the labels for the full-text articles are either 1 or 0.5, is this supposed to be the case? Training using full-text articles did not improve anything for me over base abstracts, but I assumed this would be due to the fact its not learning anything as it can just predict 1 if there are a large number of tokens.

**Comment by sujunhao** [comment]
Cool! I'm delighted to hear that BiomedNLP-PubMedBERT from Microsoft works nicely with it.

For full-text, the parsed label should be 1, 0.5, or 0. I apologize for not uploading the correct parsed full-text data to the server; we have immediately corrected this. Now you may download the right parsed file from this repo's [link](http://www.bio8.cs.hku.hk/RENET2/renet2_data_models.tar.gz) now.
For your convenience, I've also attached the revised full-text [labels.txt](https://github.com/sujunhao/RENET2/files/6781337/labels.txt) file to this comment.

Using insufficient information, I am unable to make any conclusions about the non-improvement situation of training with the full-text dataset. However, if I want to use full-text for training in an abstract model, for your reference, there are a few things I'll keep in mind:

1. Does the model digest full-text content correctly?
We may need to ensure that we fed models with accurate gene-disease pairs with the text content into the model because many gene-disease pairs in full-text are mentioned only beyond the abstract scope (i.e., certain gene-disease pairs may **only** be mentioned in or near the ending section of full-text).

2. When training on a full-text dataset, how to balance model specificity and sensitivity.
Full-text has more text information (better sensitivity for more information, but lower precision due to noise information) and a significantly higher inbalance positive/negative labeling rate in the training dataset. We may end up with a high specificity but reduced sensetivity of model if we feed all full-text pairs to training models (a majority of full-text gene-disease pairs are non-associated and non-labeled in our dataset). 

Thanks again for pointing out the empty 0 situation in the parsed file, and using our dataset, if you have any question regrading our dataset and model, feel free to let me know.

Junhao

**Comment by jambo6** [comment]
Thanks a lot for the updated labels. I'm finding another slight issue in that there are quite a lot of labels that do not correspond to gene/disease annotations. Additionally, almost all of these have label 0.

Below are the list of PMID/GeneId/DiseaseId/Labels from ft_data/labels that do not have corresponding annotations in anns.txt and thus cannot be mapped to. This could be down to a few mistakes in labelling, however it seems odd that there are so many errors in the label=0 column, and there are actually less 0 labels in the label.txt set, making this seem unlikely to be through chance. Do you have any idea what is going on here?
```
PMID: 16963499, Gene: 54487, Disease: C563466, Label: 0.0
PMID: 19059699, Gene: 4352, Disease: D001855, Label: 1.0
PMID: 19059699, Gene: 4352, Disease: D005198, Label: 0.0
PMID: 19191756, Gene: 5741, Disease: D016301, Label: 0.0
PMID: 21364575, Gene: 10178, Disease: D001943, Label: 0.0
PMID: 21388547, Gene: 10273, Disease: D009101, Label: 0.0
PMID: 21396159, Gene: 3557, Disease: C566610, Label: 0.0
PMID: 21396159, Gene: 3557, Disease: D062787, Label: 0.0
PMID: 21934688, Gene: 6302, Disease: D001943, Label: 0.0
PMID: 21983962, Gene: 54583, Disease: D005198, Label: 0.0
PMID: 22337246, Gene: 1351, Disease: D001172, Label: 0.0
PMID: 23022873, Gene: 7295, Disease: D062787, Label: 0.0
PMID: 23225887, Gene: 3553, Disease: D001281, Label: 0.0
PMID: 24080184, Gene: 6548, Disease: D003643, Label: 0.0
PMID: 24080184, Gene: 6548, Disease: D003645, Label: 1.0
PMID: 24080184, Gene: 6548, Disease: D004827, Label: 1.0
PMID: 24080184, Gene: 6548, Disease: D006130, Label: 1.0
PMID: 24509439, Gene: 2335, Disease: C537734, Label: 0.0
PMID: 24509439, Gene: 2335, Disease: D014842, Label: 0.0
PMID: 24509439, Gene: 2335, Disease: D017180, Label: 0.0
PMID: 25273894, Gene: 5243, Disease: D011535, Label: 0.0
PMID: 25926995, Gene: 5327, Disease: D016470, Label: 0.0
PMID: 25946018, Gene: 7099, Disease: D005198, Label: 0.0
PMID: 25946018, Gene: 7099, Disease: D011595, Label: 0.0
PMID: 25965909, Gene: 930, Disease: D013313, Label: 0.0
PMID: 26243863, Gene: 2064, Disease: D006816, Label: 0.0
PMID: 26410082, Gene: 5307, Disease: D003025, Label: 0.0
PMID: 26410082, Gene: 673, Disease: D015179, Label: 1.0
PMID: 26796641, Gene: 132954, Disease: D002311, Label: 0.0
PMID: 26796641, Gene: 5266, Disease: D002311, Label: 0.0
PMID: 26796641, Gene: 7015, Disease: D002311, Label: 0.0
PMID: 27222925, Gene: 5354, Disease: D004401, Label: 0.0
PMID: 27222925, Gene: 5354, Disease: D009127, Label: 0.0
PMID: 27222925, Gene: 5354, Disease: D011015, Label: 0.0
PMID: 27525306, Gene: 3082, Disease: D011535, Label: 0.0
PMID: 27642274, Gene: 2516, Disease: D003681, Label: 0.0
PMID: 27642274, Gene: 2516, Disease: D005198, Label: 0.0
PMID: 27642274, Gene: 2516, Disease: D011595, Label: 0.0
PMID: 27808126, Gene: 7295, Disease: D011125, Label: 0.0
PMID: 27880818, Gene: 6566, Disease: D005198, Label: 0.0
PMID: 27880818, Gene: 6566, Disease: D011535, Label: 0.0
PMID: 28612977, Gene: 671, Disease: D000275, Label: 0.0
PMID: 28777375, Gene: 598, Disease: C537985, Label: 0.0
PMID: 28811962, Gene: 4919, Disease: D005198, Label: 0.0
PMID: 28811962, Gene: 4919, Disease: D014842, Label: 0.0
PMID: 29393430, Gene: 79727, Disease: D030342, Label: 0.0
PMID: 29955035, Gene: 10855, Disease: C535729, Label: 0.0
PMID: 29955035, Gene: 10855, Disease: D011535, Label: 0.0
PMID: 30017315, Gene: 1194357, Disease: D053632, Label: 0.0
PMID: 30024510, Gene: 2678, Disease: D001145, Label: 0.0
PMID: 30024510, Gene: 2678, Disease: D002761, Label: 0.0
PMID: 30024510, Gene: 2678, Disease: D004066, Label: 0.0
PMID: 30024510, Gene: 2678, Disease: D009369, Label: 0.0
PMID: 30024510, Gene: 2678, Disease: D010182, Label: 0.0
```


Sorry for all the questions, this is all super helpful!

**Comment by sujunhao** [comment]
What you described is an expected situation, though it may appear weird at first. Sorry for we did not explain it clearly in our repo. But, in short, **you can safely delete all these label** when you are training and testing your models. 

These gene-disease pairs do not have gene/disease annotations and are paris that only mentioned in full text in the "Method" section (if you look at these annotations in [PTC](https://www.ncbi.nlm.nih.gov/research/pubtator/index.html), you can confirm that they are all isolated pairs from a single Method section).

We removed it because we discovered that **these labels did not help with the full-text level relation extraction problem**:

We include all full-text parts in our dataset while annotating full-text articles and training RENET2 models. However, after completing the annotation process, we discovered that our model performs better (and more reasonable way to do full-text RE) when trained on data in which a section has been removed. If you look at [our paper's Figure 3D](https://academic.oup.com/nargab/article/3/3/lqab062/6315218), the section we deleted is the Method section (which we found later, its context is generally unrelated to gene-disease relation in most of the cases) by default. 

Sorry for our poor document. We include these labels for data completeness and support our finding that the Methods section is not helpful in our analysis. We recommend to delete all these labels when you are training and testing your models. (include it when training and exclude in tesing as we did will make thing complex and unnecessary) But if you want to parse full-text data with all sections instead our default configure (which we are not recommended), you can check our example in src/nb_scripts/pre_process/preparing_ft_data.ipynb, by using `ori_tar=['TITLE', 'ABSTRACT', 'INTRO', 'METHODS', 'RESULTS', 'DISCUSS', 'CONCL']`.

Junhao

**Comment by jambo6** [comment]
Ok wonderful. Thanks again for all this. I've still had no luck improving things with the full-text data, however I will keep playing with it. 

If you are interested, I made a [fine tuned version of microsofts BioNLP model on huggingface](https://huggingface.co/jambo/BiomedNLP-PubMedBERT-base-uncased-abstract-fulltext-finetuned-renet) (I hope this is ok!). This uses only the abstracts (as is limited to 512 tokens) but achieved excellent performance on the test set (I've since reuploaded where I merge the test set into the train set).  

One thing I think is a slight issue with measuring performance on this abstract level stuff is that it is in general very easy to predict no association for genes and diseases that do not occur in the same sentence. If there is no sentence for which a given gene and disease are co-located, then with almost probability 1 the abstract is not suggesting an association. This muddies the performance metrics a bit. 

I'd quite like to have a go at manually improving the labels in the GAD dataset at some point and supplementing the two to see whether performance can significantly enhanced. Sadly there are over 5000 sentences so this may take me a while to get round to.

**Comment by sujunhao** [comment]
The results from your fine-tuned seem quite awesome and impressive. Many thanks for playing with our data. Do you mind if we put your results link on our README page? This will provide some information/reference for others when they try to apply our data at Hugging Face API and Microsoft's BiomedNLP.

For the performance problem, from my point of view, full-text and abstract models are somehow quite different as text from full-text is not informational dense (in terms of classifying a specific relationship) comparing to sentence and abstract. So there may remain some bridge to apply them together.

For the non-co-located gene and disease problem, I think it is true that most of these cases are non-associated, but there remain some cases we observed from biomedical articles that a gene and disease could be associated despite they are not co-located within any sentence from the provided text.  Though there are limited cases we found, they can happen in the original text or cause by NER errors. 
For example, in PMID: [12117686](https://www.ncbi.nlm.nih.gov/research/pubtator/index.html?view=docsum&query=12117686), the disease, Hereditary haemochromatosis, is associated with HFE (if you have a look at the context), and they are not co-located, as the authors did not explicitly discuss them within a sentence.
And in PMID: [10234502](https://www.ncbi.nlm.nih.gov/research/pubtator/index.html?view=docsum&query=10234502), the Cowden disease is associated with TEP1, but the term "CD" is not recalled by the NER, so they are also not co-located within any sentences by the view of models.


I wish you all the best if you do the manually improving process. But 5000 sentences is not a small number. (i don't know whether it is appropriate or not to writing our experience here, but this is a point we did not notice before we are doing annotation, and i'd like to write here for your information: what we learn from our manual annotation process is having a user-friendly visualization tool that showing all information (text, NER, highlighted target pair of gene and disease, a quick switch to next pairs/text) saved us a lot of effort and time when doing manual annotation, especially when doing in full-text scale. We wrote a simple python notebook script to do it, and it's work good in our work. )

Best,
JH

**Comment by jambo6** [comment]
Absolutely you can put it on the readme. 

The full-text thing is tricky. I did try stripping out everything but the sentences when the gene and disease mentions were co-located but this also did not improve performance. 

Thanks for the tip, if ever I get that far I will be sure to do something like that.

Can I confirm that you got results on the full set of PMCID texts, but have not yet applied it to the PMID abstracts?

**Comment by sujunhao** [comment]
Thanks a lot.

We also tried several strategies to filter noise text from original full-text before, like only using the target's co-located sentences, only the target's co-located paragraph, etc. Still, the results showed not improve a lot. BTW, for your information, in our full-text dataset, the text is already filtered. (We did it by a strategy called section filtering, it delete all paragraphs from full-text that did not contain any gene and disease pairs). 

And yes, we did apply RENET2 on the whole PMCID set from PMC but did not try it on PMID from PubMed. However, RENET(2019) did apply to all articles from PubMed when it was published.

> Wu, Ye, et al. "Renet: A deep learning approach for extracting gene-disease associations from literature." International Conference on Research in Computational Molecular Biology. Springer, Cham, 2019.

Best,
JH
